src/runtime/type.ts

```diff
@@ -19,9 +20,217 @@ import {CRDTTypeRecord, CRDTModel} from './crdt/crdt.js';
 import {CRDTCount} from './crdt/crdt-count.js';
 import {CRDTCollection} from './crdt/crdt-collection.js';
 import {CRDTSingleton} from './crdt/crdt-singleton.js';
+import {CRDTEntity, SingletonEntityModel, CollectionEntityModel} from './crdt/crdt-entity.js';
 import {CollectionHandle, SingletonHandle, Handle} from './storageNG/handle.js';
+import {ParticleExecutionContext} from './particle-execution-context.js';
+import {Referenceable} from './crdt/crdt-collection.js';
 
 
+export class Schema {
+  readonly names: string[];
+  // tslint:disable-next-line: no-any
+  readonly fields: Dictionary<any>;
+  description: Dictionary<string> = {};
+  isAlias: boolean;
+
+  // For convenience, primitive field types can be specified as {name: 'Type'}
+  // in `fields`; the constructor will convert these to the correct schema form.
+  // tslint:disable-next-line: no-any
+  constructor(names: string[], fields: Dictionary<any>, description?) {
+    this.names = names;
+    this.fields = {};
+    for (const [name, field] of Object.entries(fields)) {
+      if (typeof(field) === 'string') {
+        this.fields[name] = {kind: 'schema-primitive', type: field};
+      } else {
+        this.fields[name] = field;
+      }
+    }
+    if (description) {
+      description.description.forEach(desc => this.description[desc.name] = desc.pattern || desc.patterns[0]);
+    }
+  }
+
+  toLiteral() {
+    const fields = {};
+    const updateField = field => {
+      if (field.kind === 'schema-reference') {
+        const schema = field.schema;
+        return {kind: 'schema-reference', schema: {kind: schema.kind, model: schema.model.toLiteral()}};
+      } else if (field.kind === 'schema-collection') {
+        return {kind: 'schema-collection', schema: updateField(field.schema)};
+      } else {
+        return field;
+      }
+    };
+    for (const key of Object.keys(this.fields)) {
+      fields[key] = updateField(this.fields[key]);
+    }
+
+    return {names: this.names, fields, description: this.description};
+  }
+
+  static fromLiteral(data = {fields: {}, names: [], description: {}}) {
+    const fields = {};
+    const updateField = field => {
+      if (field.kind === 'schema-reference') {
+        const schema = field.schema;
+        return {kind: 'schema-reference', schema: {kind: schema.kind, model: Type.fromLiteral(schema.model)}};
+      } else if (field.kind === 'schema-collection') {
+        return {kind: 'schema-collection', schema: updateField(field.schema)};
+      } else {
+        return field;
+      }
+    };
+    for (const key of Object.keys(data.fields)) {
+      fields[key] = updateField(data.fields[key]);
+    }
+
+    const result = new Schema(data.names, fields);
+    result.description = data.description || {};
+    return result;
+  }
+
+  // TODO: This should only be an ident used in manifest parsing.
```

Comment:
Can you put my name down for this TODO?

Reply:
Done. And for good measure I put your name down for the other two unowned TODOs in this file. :-)